### PR TITLE
Clip to max_int on overflow of element_space for dynamic shape

### PR DESCRIPTION
**Pull request by CharlieL7**
* Found an error while going through code with Brian, this will make it so we throw with an descriptive error.
* Made an issue #2674 that outlines a better fix of having some way to create a dynamic shape with unknown range.

### CODE REVIEW

**Comment by pfultz2** [comment]
Would it be better to clip rather than throw an error?

**Comment by bpickrel** [comment]
I don't agree with this PR unless you apply it everywhere.  If we're going to check for 64-bit integer overflows in this case, we ought to check for them consistently in many other places in the code base.  I see that std::accumulate is used in dozens of different places.

As for the specific problem you're addressing here, I don't think it will be a pressing issue once we address the underlying issue of dynamic element spaces getting too large.  The practical limit on their size is--as far as I can tell--the amount of memory that can be allocated on the GPU, which is hardware-dependent but always far less than 2^64.  It's not clear to me that all of the memory for a dynamic element space really needs to be allocated, but apparently that's what it's trying to do right now.

I say this because I've been doing some trial and error checking of the workable limits for that Verify test I wrote for dynamic Resize op inputs.  It failed at 64x64x2kx2k.  The total element space blows up very quickly for any tensor of rank higher than 2 dimensions.

File: test/shape_test.cpp [hunk 1]
```diff
@@ -160,6 +160,34 @@ TEST_CASE(test_shape_dynamic_compares)
...
+    migraphx::shape s{migraphx::shape::float_type, {{0, max_val}, {0, max_val}}};
+    EXPECT(s.element_space() == max_val);
+}
+
+TEST_CASE(dynamic_shape_element_space_overflow1)
+{
+    std::size_t max_val   = std::numeric_limits<std::size_t>::max();
+    std::size_t large_val = max_val / 10;
+    migraphx::shape s{migraphx::shape::float_type, {{0, large_val}, {0, large_val}}};
+    EXPECT(s.element_space() == max_val);
```
Comment (TedThemistokleous) [inline test/shape_test.cpp:181]: If we overflow do we not expect to error out/ signal an overflow as a warning/error?

File: src/shape.cpp [hunk 2]
```diff
@@ -123,7 +123,17 @@ struct shape_impl
...
             auto maxes = max_lens();
-            return std::accumulate(maxes.begin(), maxes.end(), std::size_t{1}, std::multiplies<>());
+            std::size_t max_val = std::numeric_limits<std::size_t>::max();
+
+            return std::accumulate(
+                maxes.begin(), maxes.end(), std::size_t{1}, [&](std::size_t x, std::size_t y) {
+                    // overflow check and clip
+                    if(x != 0 and y > max_val / x)
+                    {
+                        return max_val;
```
Comment (TedThemistokleous) [inline src/shape.cpp:133]: add a warning about clipping that we can turn on in debug builds

**Review by TedThemistokleous — COMMENTED** [review]
I think its better to clip but then add in a warning/error during debug builds to show that's happened. This sounds like a good topic to go over for Monday

File: src/shape.cpp [hunk 3]
```diff
@@ -123,7 +123,17 @@ struct shape_impl
         if(not m_dyn_dims.empty())
         {
             auto maxes = max_lens();
-            return std::accumulate(maxes.begin(), maxes.end(), std::size_t{1}, std::multiplies<>());
+            std::size_t max_val = std::numeric_limits<std::size_t>::max();
```
Comment (pfultz2) [inline src/shape.cpp:126]: This can be declared as a constant.

**Review by pfultz2 — APPROVED** [review]

**Comment by pfultz2** [comment]
> As an example there is this MIGRAPHX_ASSERT in src/targets/gpu/kernels/include/migraphx/kernels/gemm_batcher.hpp

That is using a static shape not a dynamic shape.

**Review by TedThemistokleous — APPROVED** [review]

File: src/shape.cpp (hunk 2 above)
Comment (umangyadav) [reply src/shape.cpp:133]: Warning would be good to have

**Pull request merged**